### ISSUE TITLE
WP-CLI Integration

### ISSUE THREAD

**Issue by twiginteractive** [szepeviktor/fix-w3tc#158]
Hi - I see that W3TC has commands integrated with WP-CLI (for flushing cache etc.) but I don't see an option to kick off the caching priming function to warm the cache via iterating through the sitemap.

Is this something controlled by the W3TC plugin (i.e. this community version could add features to the WP-CLI integration list) OR is it controlled by WP-CLI, and we would need to submit new features to them for inclusion in their next release? Sorry for the noob question but I've not exposed features to WP-CLI and their website isn't clear on the topic.

Thanks!

**Comment by charlesLF** [comment]
Hey @twiginteractive,

Nice idea.

That's a feature that should be added to the W3TC plugin.
I will take a look at it when I get some time.

**Comment by twiginteractive** [comment]
@charlesLF Sweet! I look forward to seeing that feature make it in. Love the community version of this plugin.

**Comment by amiga-500** [comment]
Hi @twiginteractive  ... adding this feature would be trivial but i am curious...because priming is dependent on the priming's _limit_ and _interval_ settings as defined on the _Page Cache > Cache Preload_ page i assume you would want to also have the ability (via CLI) to also provide your own _limit_ and _interval_ values that supersede the ones defined in the admin page.  Correct?  And going further, when not providing these CLI subcommands (limit and/or interval) it would default to what was given in the admin page?  Or would it just use default values shown under the w3tc CLI help page?

Are there any other CLI commands on your wishlist?

Regards
Kimberly

**Comment by twiginteractive** [comment]
@amiga-500 (great handle BTW - i had one too!)

Thank you for the detailed and quick reply - probably the best response to a feature request I've seen on a WP plugin. To your Qs:

My use-case is probably a little 'edge' - I'm running WP as a 2-tier environment (Dev -> Stage and Production) on AWS Elastic Beanstalk, and as such the production env is made up of several boxes, autoscaled by ELB. So when I want to purge (and prime) the cache, I need to activate these functions for each node in the cluster. I used to have to do with this local shell scripts - but with WP-CLI, now I can do remote management (with the appropriate pre-ssh commands being run)

For this particular request, I need to kick off the priming function on-demand after my publishing script (which pushes content and files from Dev -> Stage | Prod ) so that each node happily starts to create those cached pages ready for the public to view. The site in question has ~50,000 pages of content (driven from another data source) that get updated daily - so the flow is to flush + prime the cache once each day in the early hours.

I've gotten the cache flushing thing done via WP-CLI (and I have already set the priming parameters in the UI as you mention - interval + batch). So, first level of effort would just be to initiate the priming to use values stored in W3TC prefs.

However, now you mention it - it would be awesome to be able to pass further params to the `wp w3-total-cache prime <ARGS>` etc. to control the interval and batch; but I would see that as bonus vs. required.

I always disable the native WP_CRON flag, and instead use wget as part of a real server cronjob to ping wp-cron.php every minute. That way I know that at least once a minute, all WP crons are being fired. This is working fine so no worries on scheduling.

Since I would always be flushing the cache first, the cache would be expired so priming should create all pages anew. 

I can't think of any other CLI commands on my wish list.

Thanks!!
Jonathan

**Comment by amiga-500** [comment]
Hi @twiginteractive 

> @amiga-500 (great handle BTW - i had one too!)

LOL!  Awesome to see someone get this reference ;-) 

Just a heads-up.  I got the prime feature working in WP-CLI.  I am just doing some coded housekeeping and final debugging before pushing the new feature in.  Seems to all work fine but i don't want to rush it.

The syntax will be:

`wp w3-totalcache prime [--interval=<seconds>] [--batch=<#>] [--sitemap=<url>]`

All the args are optional as you can tell and will use the values defined in w3tc's prefs when missing.   I also decided to add a _stop_ arg just in case you needed to halt the active priming process:

`wp w3-totalcache prime stop`

Cheers
Kimberly

**Comment by twiginteractive** [comment]
Nice one @amiga-500 ! You work faster than autofire in Xenon II :)

I look forward to testing this out, and will give feedback as I do. Cheers!

**Comment by amiga-500** [comment]
Hi again @twiginteractive !  Just wanted to let you know i've merged the new wp-cli prime ability into master.  So if you want to test it out you can either grab the latest [master](https://github.com/szepeviktor/fix-w3tc/archive/master.zip) or alternatively grab the following specific new files and update your corresponding side:

[w3-total-cache/lib/W3/Cli.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/master/lib/W3/Cli.php)
[w3-total-cache/lib/W3/PgCacheAdminEnvironment.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/master/lib/W3/PgCacheAdminEnvironment.php)
[w3-total-cache/lib/W3/Plugin/PgCache.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/master/lib/W3/Plugin/PgCache.php)
[w3-total-cache/lib/W3/Plugin/PgCacheAdmin.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/master/lib/W3/Plugin/PgCacheAdmin.php)

If you find issues just let me know. :-)

Cheers!

**Comment by twiginteractive** [comment]
Hi @amiga-500  - I've updated master on my Dev box and I'm testing this new feature out now. Here are my comment so far:

- As I understand it from inspection of httpd logs and using WP-CRONTROL, W3TC creates a `w3_pgcache_prime` event (with no parameters), with a recurrence based on the plugin settings. This event is (re)created if needed when you hit Save All Settings from the plugin.

This job iterates through the sitemap and collects the URLs to process. When this is done, it creates **another** `w3_pgcache_prime` event (this time with parameters pulled from the preferences) and this job actually parses the URL list and creates the cached files.

- In the W3TC plugin, I have the prefs set batch=50000, interval=1day, sitemap=/sitemap_index.xml.
Before this plugin update, the URL-processing job had params [50000].
After plugin update, it has params [50000, -1, -1, ''].
Is this intentional? I'm not sure if passing -1 and empty string for the sitemap will mess anything up? (i.e. if providing these defaults will be the same as taking function defaults)

- If I issue command `wp w3-total-cache prime stop` it will kill not only the running event (the one with the params) - but also the parent scheduled event. Meaning that the cache priming will never run again unless/until I hit Save All Settings to recreate the schedule. Is it possible for the `stop` command to only kill the parsing job (the one with params), and leave the parent scheduled to run next time the interval hits?

- I see that the script will not terminate until all the URLs have been processed (which for my site will take several hours with 50K URLs). I'm guessing I should `nice nohup` the `wp w3-total-cache prime` command if I want to use this as part of a larger deployment script?

Most importantly, THANK YOU for adding this feature! Let me know what you think about my Qs.

Cheers!

**Comment by amiga-500** [comment]
Hi @twiginteractive 

Firstly, much appreciated on the awesome feedback!  I've inlined my answers below.

> _As I understand it from inspection of httpd logs and using WP-CRONTROL, W3TC creates a `w3_pgcache_prime` event (with no parameters), with a recurrence based on the plugin settings. This event is (re)created if needed when you hit Save All Settings from the plugin. This job iterates through the sitemap and collects the URLs to process. When this is done, it creates **another** `w3_pgcache_prime` event (this time with parameters pulled from the preferences) and this job actually parses the URL list and creates the cached files._

You almost have it right.  When you "_Save All Settings_" and "_Automatically prime the page cache_" was originally unchecked but is now checked W3TC creates a scheduled event that runs immediately and processes all urls up to the batch limit (starting from the very beginning of the sitemap).  If there are more urls to process than the batch limit allows then it is only then it creates another (2nd) event to handle the next batch (which runs after "_Update Interval_" seconds has passed).  This 2nd event is a 1-timer -- it won't repeat.  The 1st event (the original one) though, is a repeater and will repeat every "_Update Interval_" seconds and start the priming process from the very beginning of the sitemap (assuming no existing 2nd event -- for batch handling -- is running).  If that 2nd event, after it runs and processes it's assigned batch, still has more urls to process, it creates a non-repeating 1-time event to handle that next batch.  So, when you use WP-CRONTROL you can potentially see up-to two w3_pgcache_prime events.  Hope that isn't too confusing to understand.

What i found surprising was that by default W3TC runs its priming process immediately when you click the "_Save All Settings_" (assuming the "_Automatically prime the page cache_" was originally saved as unchecked but is now checked on) .  I had originally thought it would wait "_Update Interval_" seconds before beginning with the 1st url within the sitemap.

> _In the W3TC plugin, I have the prefs set batch=50000, interval=1day, sitemap=/sitemap_index.xml.
> Before this plugin update, the URL-processing job had params [50000].
> After plugin update, it has params [50000, -1, -1, ''].
> Is this intentional? I'm not sure if passing -1 and empty string for the sitemap will mess anything up? (i.e. if providing these defaults will be the same as taking function defaults)_

Yes that [50000,-1,-1,'']  is normal and intentional.  The extra parameters were added to help feed in CLI-defined parameters, if any were provided.  If none were given the defaults of "-1" or "" (empty string) tells the prime function to use the W3TC preferences found under _Cache Control_ instead.

> _If I issue command `wp w3-total-cache prime stop` it will kill not only the running event (the one with the params) - but also the parent scheduled event. Meaning that the cache priming will never run again unless/until I hit Save All Settings to recreate the schedule. Is it possible for the `stop` command to only kill the parsing job (the one with params), and leave the parent scheduled to run next time the interval hits?_

Right, right!  I forgot about that.  Whoops!  I had originally figured the priming would either be done via CLI or via the "_Automatically prime the page cache_" option, but not both.  My mistake.  I had incorrectly figured the "_prime stop_" should just kill all active priming events since it was the only one who was assigned to prime.

That being said, i will rewrite this so that for CLI purposes it will have its own priming event called: "_w3_pgcache_prime_cli_" so that it can do its thing independent of the Automatic prime event (aka "_w3_pgcache_prime_").

> _It seems that `prime stop` won't actually stop the processing of the current batch of URLs, but rather prevent further batches from being executed. This is probably intentional, but as I was doing 1 big batch daily, the stop command doesn't actually stop the processing. I could change the settings to use a smaller batch and a smaller window I guess?_

Ahhh..that's a good point.  I forgot about that.  Yea because when the script is running (and hence, priming) any stop request will only halt future batch requests, not the ongoing one.  So it seems one would need to kill an active wp-cli PID that is doing the priming -- that will stop the priming promptly.  The easiest approach would be to write to disk the PID when it starts priming and if a stop request is made it checks for this file (to get the original running PID) and kill the process.  And the prime stop would also still remove the prime event, originally initiated by cli, (if one exists).

> _I see that the script will not terminate until all the URLs have been processed (which for my site will take several hours with 50K URLs). I'm guessing I should `nice nohup` the `wp w3-total-cache prime` command if I want to use this as part of a larger deployment script? Or would it be better if the WP-CLI command itself `nohup`d the process and just returned message "Priming has been triggered" instead of waiting to report "Priming has been completed" ?_

It would be nice if the WP-CLI command itself nohup'd and returned back a  "_Priming has been triggered_" response.  I forgot that when each batch is large it wont return back immediately.  I will set up the priming process in W3TC (for cli only) so that when a request to prime is started it will trigger a wp-cron event that begins the actual priming so that it will return immediately with a "Priming has been started" response.  As mentioned earlier, the reason it halts is because W3TC naturally begins the prime process when the event is created, and it does so by iterating through a blocking loop.  Because i will be using my own w3_pgcache_prime_cli event and function logic in the new update today it should solve the problem and return back immediately. ;-)

When i get home i will make an update and let you know.

Thanks again for the feedback!

Cheers!
Kimberly

**Comment by amiga-500** [comment]
Sorry for the delay @twiginteractive ... ended up doing some shopping for the day.  Anyway, i've just committed a new WP-CLI revision for ya to test.  If i've interpreted your feedback correctly this new revision should resolves all your issues.  Definitely let me know.  Happy to keep tweaking this until its right ;-)

I decided to not merge this revision yet with the master branch.  So at present it's currently separate within the _wp-cli-revision_ branch temporarily.  This way we can continue to make any further tweaks if necessary to this revision get it working just right before I do the offiical merge.

So to test things out in this new revision you can either grab the entire zip from [wp-cli-revision](https://github.com/szepeviktor/fix-w3tc/archive/wp-cli-revision.zip) or alternatively grab the following specific new files and update your corresponding side:

[w3-total-cache/inc/define.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/b2431814aa35d297c5d48cbe043716fe9eb6d5fc/inc/define.php)
[w3-total-cache/lib/W3/Cli.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/b2431814aa35d297c5d48cbe043716fe9eb6d5fc/lib/W3/Cli.php)
[w3-total-cache/lib/W3/PgCacheAdminEnvironment.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/b2431814aa35d297c5d48cbe043716fe9eb6d5fc/lib/W3/PgCacheAdminEnvironment.php)
[w3-total-cache/lib/W3/Plugin/PgCache.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/b2431814aa35d297c5d48cbe043716fe9eb6d5fc/lib/W3/Plugin/PgCache.php)
[w3-total-cache/lib/W3/Plugin/PgCacheAdmin.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/b2431814aa35d297c5d48cbe043716fe9eb6d5fc/lib/W3/Plugin/PgCacheAdmin.php)

And as always much appreciated for the past and future feedback and any questions you might want to know.

Cheers
Kimberly

**Comment by amiga-500** [comment]
Hi @twiginteractive ... haven't heard back from you in a while.  :(

**Comment by twiginteractive** [comment]
@amiga-500 
Hey Kimberly! Sorry for the radio silence - Halloween weekend always (and quite rightly) derails my work train! I'm back in the office today though, and have read your very detailed posts above. I'll grab the latest from the new branch and test it out and get you feedback probably by tomorrow.

I have to say I'm most impressed with your diligence & responsiveness for a free community plugin - do you do contract WordPress work??!

**Comment by twiginteractive** [comment]
Hi again @amiga-500 

I updated my install from your new branch and everything looks great. I was able to trigger and stop the priming function for both local and remote hosts using WP-CLI. Was there any other specific testing you would like me to do to help you get this merged into master?

I'll continue to use it in my publishing workflow and comment on this thread with any feedback.

Thanks again!

**Comment by amiga-500** [comment]
Hi Jonathan!

> _I updated my install from your new branch and everything looks great.   I was able to trigger and stop the priming function for both local and remote hosts using WP-CLI._

Yay! Fantastic to here!

> _Was there any other specific testing you would like me to do to help you get this merged into master?_

Nope....that's all i needed to know.

> _I have to say I'm most impressed with your diligence & responsiveness for a free community plugin - do you do contract WordPress work? (If this question is too off-topic, no problem at all!)_

Hey, much appreciated for the compliment ;-) !  And nah i just do this for the fun of it on my free time.

Once again, thanks very much for the feedback and have a good one!

Kimberly
_Amiga forever!_

**Comment by twiginteractive** [comment]
Hi @amiga-500 - me again :)

I've been using this feature a lot today in my deployment testing and it works great. One thing I did noticed through (tail-ing the logs in real-time) is that the manual priming function runs through the sitemap.xml twice before it starts actually hitting the pages.

This may be intentional - but I wanted to bring it up. 
So the process I see is:

* Issue the `wp w3-total-cache prime` command
* Logs show the main parent sitemap.xml is hit, then each child sitemap xml (we have 60 of them) is processed
* Once all the sitemaps have been processed, THEN the new w3_pgcache_prime_cl event is created
* After a short delay, this event fires, and hits the sitemap.xml (and then child sitemaps) AGAIN before actually then hitting the pages themselves

Thought you would like the feedback.

Thanks Kimberly!

**Comment by amiga-500** [comment]
Hi @twiginteractive ....thanks for the info!

Yup that is exactly what is happening.  There's a wordpress-specific event necessity for that which i dont like but as i write this i just realized i should be able to do a quick modification to avoid that and have it just call sitemap.xml once.  It would involve a read of sitemap.xml and passing the urls collection held within this sitemap (main + children) as a parameter of the event so that all future events no longer need to hit the sitemap to know what urls to process).  This would save on some CPU cycles.   I'll go play around and get back to ya with a tweak later tonight.  Should be fun. :)

Anyway, i'm off to go see Doctor Strange in theatres.

Kimberly

**Comment by twiginteractive** [comment]
Great sleuthing again @amiga-500 !

I'm still using my "one daily big prime event of 50K pages" but I can see that for people that wanted to chunk it up into smaller pieces, avoiding the repeated parsing of the sitemaps would be sweet. I'm happy to test + report once you have a new version!

(Jealous re: Dr S. I'd rather see some Crabberbatch than working on WP updates as I am right now! Can't really complain though since this whole long weekend was Halloween nerd-fun)

**Comment by amiga-500** [comment]
Oh hi Jonathan!  Didn't realize you are still around.  Hey, funny enough i think i am done updating it.  I haven't checked it out but it was so simple a tweak that it looks like its ready to go.  I will post some links for ya to update on your side to test it out..just 3 files changes.

As for the 10,000 URLS...it will be passed as 1 parameter within an array structure of the event so memory would be the only concern.  Let's say each url is 80 bytes large and so at 10,000 entries that means it would use 781KB.  So it's quite tiny for its memory footprint.

I'll post some files for ya shortly.

**Comment by amiga-500** [comment]
The only issue i can think of for the 10,000 URLs then would be if you try to view the cron within say WP-Crontrol plugin inside WP you will see some 10,000 URLs double-quoted as a single parameter because that plugin is trying to be helpful by displaying the full list of what is inside that parameter (no truncation for display here)...internally its passed as a single array parameter not as 10,000 parameters.

Here are the update files (or here is the full package: [wp-cli-revision.zip](https://github.com/szepeviktor/fix-w3tc/archive/wp-cli-revision.zip)):

[/lib/W3/Cli.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/cc5ff941230d44f11c10eed40e4a51ab18eaf24e/lib/W3/Cli.php)
[/lib/W3/Plugin/PgCache.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/cc5ff941230d44f11c10eed40e4a51ab18eaf24e/lib/W3/Plugin/PgCache.php)
[/lib/W3/Plugin/PgCacheAdmin.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/cc5ff941230d44f11c10eed40e4a51ab18eaf24e/lib/W3/Plugin/PgCacheAdmin.php)


Ok i am officially gone now :p   Let me know how it goes and i will do changes and such when i get back from the movie.

**Comment by twiginteractive** [comment]
Heh I guess that was my concern since I was looking at the params via that plugin :|
I have about 55,000 URLs in this particular client site soooooo that could be interesting seeing that explode the DOM for the crontrol view!

I'm done for the night but will def check this out tomorrow and let you know! Enjoy the flick.

**Comment by twiginteractive** [comment]
Hey @amiga-500  - 

(FYI the links to the package https://github.com/szepeviktor/fix-w3tc/archive/wp-cli-revision.zip gives a 404)

I updated the individual files - I think we might need to revert back because the output from a `prime` command borked. It spews several screens of serialized data and then the PHP error below

`PHP Warning:  Error while sending QUERY packet. PID=26640 in /var/app/current/wp-includes/wp-db.php on line 1868
Warning: Error while sending QUERY packet. PID=26640 in /var/app/current/wp-includes/wp-db.php on line 1868`

As mentioned, I gotta lot of URLs :P

**Comment by amiga-500** [comment]
Good morning @twiginteractive ,

Ok i made some modifications this morning ...

1. It now only checks the sitemap (and its children) once.

1. The idea about passing the urls as a parameter through subsequent events has been dropped.

1. On the command-line (or via w3tc prefs' _Pages per interval_ box ) i've added the ability so it can accept a batch limit size of 0 (ie. _--batch=0_) which tells it to cache all pages within the sitemap without the use of a batch.  At least that way you don't have to set the batch size to a monster size just to get it to do your entire sitemap in 1 pass.  So, in your case you could do something like:

 wp total-cache prime --batch=0 --interval=0 --sitemap=http://domain.com/sitemap.xml
 

The changed files for download are listed below.  Alternatively, the full package with the changes are: [wp-cli-revision-2.zip](https://github.com/szepeviktor/fix-w3tc/archive/wp-cli-revision-2.zip).

[/inc/define.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/9b3115ea11463a1ae596d6af91580d8805ecd09a/inc/define.php)
[/lib/W3/Cli.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/9b3115ea11463a1ae596d6af91580d8805ecd09a/lib/W3/Cli.php)
[/lib/W3/Plugin/PgCache.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/9b3115ea11463a1ae596d6af91580d8805ecd09a/lib/W3/Plugin/PgCache.php)
[/lib/W3/Plugin/PgCacheAdmin.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/9b3115ea11463a1ae596d6af91580d8805ecd09a/lib/W3/Plugin/PgCacheAdmin.php)

Let me know how it goes.

Thanks
Kimberly

**Comment by amiga-500** [comment]
> I updated the individual files - I think we might need to revert back because the output from a prime command borked. It spews several screens of serialized data and then the PHP error below

Oops sorry about that @twiginteractive   ... i was updating those files and deleted the ones you were attempting to download (the ones i posted last night).  Yea try the ones i just posted right now (the reply above this one).

**Comment by twiginteractive** [comment]
Ha! Guess we were both keen to test this out :)

Ok - working with the new set now, and I love the idea of a 0-batch. Maybe you can add text on the _Pages per Interval_ box on the prefs page to say "Set 0 for no limit" so that other folks will benefit?

Good news: your manual priming function works nicely: sitemaps run through once then BOOM pages start getting made :+1: 

Bad news: it seems the native W3 scheduled prime no longer works. As a test, I set it to run 100 pages every 5 mins. I watched the `w3_pgcache_prime` event get scheduled (and reported as running "now" when the time came) but watching the HTTPD logs - no action and no pages created.

Could it be that the native function is failing with the new updates?

**Comment by amiga-500** [comment]
> _Maybe you can add text on the Pages per Interval box on the prefs page to say "Set 0 for no limit" so that other folks will benefit?_

Good idea!  Will do.

> _Bad news: it seems the native W3 scheduled prime no longer works. As a test, I set it to run 100 pages every 5 mins. I watched the w3_pgcache_prime event get scheduled (and reported as running "now" when the time came) but watching the HTTPD logs - no action and no pages created._

Ahh..that's interesting to know.  I will go snoop around and get back to ya.

Thanks again for the info.

**Comment by amiga-500** [comment]
Eeek...found the problem...yup my mistake..sorry.  Fixing now.

**Comment by twiginteractive** [comment]
Are you sick of this thread yet? :smile: 

So as Columbo would say - 'one more question...'
I noticed that once I have executed `prime` then let it run ( or stopped it via `prime stop`) the next time I run `prime` it doesn't hit the sitemap.xmls at all (but it does start executing the pages after a short delay)

Does this mean the sitemap.xml results are being cached somewhere? Only reason I ask is that our sitemaps change daily (around 2am) so would want to be sure that after a completed run, the next time it ran it would get the latest set of URLs from the sitemaps

**Comment by amiga-500** [comment]
> _Are you sick of this thread yet? :smile:_ 

Hahahaha...never!  It's definitely fun!  :stuck_out_tongue_winking_eye:

> _I noticed that once I have executed prime then let it run ( or stopped it via prime stop) the next time I run prime it doesn't hit the sitemap.xmls at all (but it does start executing the pages after a short delay)_

That's strange.  I definitely don't cache it.  When the prime finishes (or is stopped) and then is started up again it should be attempting to retrieve the sitemap.xml data again.  No caching.  Do you have a Varnish cache server running on your side? (it's possible the server is serving back cache information since the sitemap is being retrieved via url.  I might have to use a postfixed timestamp to fool the server into feeding back a fresh page).  But before considering that I will jump into w3tc's sitemap reader code to see what its doing.

**Comment by twiginteractive** [comment]
Nope - no Varnish or reverse proxy stuff here. I tried flushing all the caches via WP-CLI and re-priming - same deal. It hits the pages but I don't see fresh hits to *.xml.

Also - if I deactivate the plugin, it doesn't clear any in-process `w3_pgcache_prime_cli` priming events - maybe add that whilst you are in there? :)

**Comment by amiga-500** [comment]
Ooops never mind...i found the problem.  LOL!  You're quite the debugger.  I will have new files for ya in  a sec to try.

**Comment by twiginteractive** [comment]
Ha! Awesome. You're quite the Dev, so I guess we make a good pair!

**Comment by amiga-500** [comment]
Hi.  sorry for the delay.  Just on the phone with family so its delaying things.  I will have the files up eventually.

**Comment by twiginteractive** [comment]
LOL - priorities! *jk*

**Comment by amiga-500** [comment]
> _LOL - priorities! jk_

ROFL!  Hilarious!

Ok just pushed in a new version for ya to try: [wp-cli-revision-3.zip](https://github.com/szepeviktor/fix-w3tc/archive/wp-cli-revision-3.zip)

**_or_** just the specific files:

[inc/define.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/1953d2d8ed6059c02d9e3deacfd83f41c053d409/inc/define.php)
[inc/options/pgcache.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/1953d2d8ed6059c02d9e3deacfd83f41c053d409/inc/options/pgcache.php)
[lib/W3/Cli.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/1953d2d8ed6059c02d9e3deacfd83f41c053d409/lib/W3/Cli.php)
[lib/W3/Root.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/1953d2d8ed6059c02d9e3deacfd83f41c053d409/lib/W3/Root.php)
[lib/W3/Plugin/PgCache.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/1953d2d8ed6059c02d9e3deacfd83f41c053d409/lib/W3/Plugin/PgCache.php)
[lib/W3/Plugin/PgCacheAdmin.php](https://raw.githubusercontent.com/szepeviktor/fix-w3tc/1953d2d8ed6059c02d9e3deacfd83f41c053d409/lib/W3/Plugin/PgCacheAdmin.php)

**Comment by twiginteractive** [comment]
Sweet - thanks! In the middle of something right now (gotta pay the bills :wink:) but will check it out today & hit you back.

**Comment by twiginteractive** [comment]
Alrighty - I'm checking out v3 and it's looking good. The native W3 job runs on schedule, and your fn hits the sitemap files each time it is kicked off. 

So - here's something for you to chew on :)
If the native W3 job is running, I can still start a manual prime (because your fn using that custom event). This means 2 sets of fn are creating cached pages. Not necessarily a bad thing, but I was going to suggest that your script checks if *any* priming is currently running. However I'm not sure that's possible when batch=0 because the W3 event starts off the process with no params then terminates. Unless you could scan the PIDs for whatever command identifies it as W3 priming?

I think this is really getting into the weeds / edge-cases but you know I'm Mr Debug! In my case, I'm actually going to disable the "auto prime" in the W3 prefs, and instead set a server cron to prime the cache after a publish OR daily. That works better for me anyway since I can control the time of day.

Ok - think that's all the feedback I have. Awesome work Kimberly!

**Comment by amiga-500** [comment]
Hi @twiginteractive  ... sorry for the delay ... catching up :running:

> _If the native W3 job is running, I can still start a manual prime (because your fn using that custom event). This means 2 sets of fn are creating cached pages. Not necessarily a bad thing, but I was going to suggest that your script checks if any priming is currently running. However I'm not sure that's possible when batch=0 because the W3 event starts off the process with no params then terminates. Unless you could scan the PIDs for whatever command identifies it as W3 priming?_

That shouldnt be a problem.  Interestingly, when i first made this prime feature it actually did check to see if the native priming was running.  It was a bit rudimentary in its check.  For whatever reason...i think i misinterpreted something you said earlier so i dropped the feature and allowed for both to run independently going as far as changing the event name -- they use to share the same name w3_pgcache_prime.  So putting back an ability to make sure W3's native priming isn't running won't be difficult.  I think the hard part was knowing what it should do in different use cases....i can't think of them all at the moment so i will post replies as i put back the feature:

1. If native is running and a CLI request to prime is started, should the CLI gracefully fail saying that the native priming is already started?  Or should the CLI's prime request take precedence and kill the active native prime process before starting its own prime (from the beginning)? If you were to say, flush the cache while the native priming is going and then start the CLI prime which notified you that native priming is already started then the native priming could still be priming let's say midway or at the tail end of the cache and so when the native was done there could potentially be half the cache missing.

1. Same as question 1 but in reverse: If CLI request is already started but native prime wants to start.

2. If the native isn't running but is queued to run  (meaning, the "Automatic Prime the Page Cache" checkbox is activated) let's say 1 hour from now, should a request to CLI prime feature kill this upcoming native prime event?

Thanks again

**Comment by twiginteractive** [comment]
Heh - seeing your detailed explanation, I actually think running concurrently isn't a bad thing. I think we could be opening many cans of worms having manual priming cancel out the native functions. 

I think I would prefer knowing that whatever I enqueue for the CLI prime is going to work fully every time. I watched the creation of the pages concurrently and it didn't cause any problems.

**Comment by amiga-500** [comment]
Ahhh ok.  So then i will not put in this check feature in CLI prime then.  If, however, there is anything else you need (or wish it could do) for CLI prime just let me know.

Cheers
Kimberly
